File: modules/utils/DatabaseService.py
import sqlite3
from pathlib import Path
from typing import List, Dict, Any, Optional
from interfaces import ISingleton
from paths import path_resolver


class DatabaseService(ISingleton):
    _instance = None
    
    def __init__(self):
        if DatabaseService._instance is not None:
            raise Exception("DatabaseService уже существует. Используйте getInstance()")
        
        self.db_path: Optional[Path] = None
        self.connection: Optional[sqlite3.Connection] = None

        self.connection = sqlite3.connect(f"{path_resolver['database_path']}/apps.db", check_same_thread=False)
        self.connection.row_factory = sqlite3.Row
    
    @classmethod
    def getInstance(cls) -> 'DatabaseService':
        if cls._instance is None:
            cls._instance = cls()
        return cls._instance
    
    
    def get_all_apps(self) -> List[Dict[str, Any]]:
        try:
            if not self.connection:
                print("[DatabaseService] Нет соединения с базой данных")
                return []
                
            cursor = self.connection.cursor()
            
            query = """
                SELECT 
                    a.id,
                    a.name,
                    a.executable_path,
                    a.icon_path,
                    a.file_size,
                    a.file_type,
                    a.last_modified,
                    a.launch_count,
                    a.last_launched,
                    a.is_favorite,
                    a.created_at,
                    p.id as path_id,
                    p.path as folder_path,
                    p.name as folder_name
                FROM apps a
                JOIN app_paths p ON a.path_id = p.id
                WHERE p.is_active = 1
                ORDER BY a.name
            """
            
            cursor.execute(query)
            rows = cursor.fetchall()
            
            apps = []
            for row in rows:
                app = {
                    'id': row['id'],
                    'name': row['name'],
                    'executable_path': row['executable_path'],
                    'icon_path': row['icon_path'],
                    'file_size': row['file_size'],
                    'file_type': row['file_type'],
                    'last_modified': row['last_modified'],
                    'launch_count': row['launch_count'],
                    'last_launched': row['last_launched'],
                    'is_favorite': bool(row['is_favorite']),
                    'created_at': row['created_at'],
                    'path_id': row['path_id'],
                    'folder_path': row['folder_path'],
                    'folder_name': row['folder_name']
                }
                apps.append(app)
            
            return apps
            
        except Exception as error:
            print(f"[DatabaseService] Ошибка получения приложений: {error}")
            return []
    
    def get_apps_by_name(self, name_filter: str) -> List[Dict[str, Any]]:
        try:
            if not self.connection:
                print("[DatabaseService] Нет соединения с базой данных")
                return []
                
            cursor = self.connection.cursor()
            
            query = """
                SELECT 
                    a.id,
                    a.name,
                    a.executable_path,
                    a.icon_path,
                    a.file_size,
                    a.file_type,
                    a.last_modified,
                    a.launch_count,
                    a.last_launched,
                    a.is_favorite,
                    a.created_at,
                    p.id as path_id,
                    p.path as folder_path,
                    p.name as folder_name
                FROM apps a
                JOIN app_paths p ON a.path_id = p.id
                WHERE p.is_active = 1 AND a.name LIKE ?
                ORDER BY a.launch_count DESC, a.name
            """
            
            cursor.execute(query, (f'%{name_filter}%',))
            rows = cursor.fetchall()
            
            apps = []
            for row in rows:
                app = {
                    'id': row['id'],
                    'name': row['name'],
                    'executable_path': row['executable_path'],
                    'icon_path': row['icon_path'],
                    'file_size': row['file_size'],
                    'file_type': row['file_type'],
                    'last_modified': row['last_modified'],
                    'launch_count': row['launch_count'],
                    'last_launched': row['last_launched'],
                    'is_favorite': bool(row['is_favorite']),
                    'created_at': row['created_at'],
                    'path_id': row['path_id'],
                    'folder_path': row['folder_path'],
                    'folder_name': row['folder_name']
                }
                apps.append(app)
            
            return apps
            
        except Exception as error:
            print(f"[DatabaseService] Ошибка фильтрации приложений: {error}")
            return []
    
    def get_apps_by_path(self, path_id: int) -> List[Dict[str, Any]]:
        try:
            if not self.connection:
                print("[DatabaseService] Нет соединения с базой данных")
                return []
                
            cursor = self.connection.cursor()
            
            query = """
                SELECT 
                    a.id,
                    a.name,
                    a.executable_path,
                    a.icon_path,
                    a.file_size,
                    a.file_type,
                    a.last_modified,
                    a.launch_count,
                    a.last_launched,
                    a.is_favorite,
                    a.created_at,
                    p.id as path_id,
                    p.path as folder_path,
                    p.name as folder_name
                FROM apps a
                JOIN app_paths p ON a.path_id = p.id
                WHERE p.is_active = 1 AND a.path_id = ?
                ORDER BY a.name
            """
            
            cursor.execute(query, (path_id,))
            rows = cursor.fetchall()
            
            apps = []
            for row in rows:
                app = {
                    'id': row['id'],
                    'name': row['name'],
                    'executable_path': row['executable_path'],
                    'icon_path': row['icon_path'],
                    'file_size': row['file_size'],
                    'file_type': row['file_type'],
                    'last_modified': row['last_modified'],
                    'launch_count': row['launch_count'],
                    'last_launched': row['last_launched'],
                    'is_favorite': bool(row['is_favorite']),
                    'created_at': row['created_at'],
                    'path_id': row['path_id'],
                    'folder_path': row['folder_path'],
                    'folder_name': row['folder_name']
                }
                apps.append(app)
            
            return apps
            
        except Exception as error:
            print(f"[DatabaseService] Ошибка получения приложений по пути: {error}")
            return []
    
    def get_favorite_apps(self) -> List[Dict[str, Any]]:
        try:
            if not self.connection:
                print("[DatabaseService] Нет соединения с базой данных")
                return []
                
            cursor = self.connection.cursor()
            
            query = """
                SELECT 
                    a.id,
                    a.name,
                    a.executable_path,
                    a.icon_path,
                    a.file_size,
                    a.file_type,
                    a.last_modified,
                    a.launch_count,
                    a.last_launched,
                    a.is_favorite,
                    a.created_at,
                    p.id as path_id,
                    p.path as folder_path,
                    p.name as folder_name
                FROM apps a
                JOIN app_paths p ON a.path_id = p.id
                WHERE p.is_active = 1 AND a.is_favorite = 1
                ORDER BY a.launch_count DESC, a.name
            """
            
            cursor.execute(query)
            rows = cursor.fetchall()
            
            apps = []
            for row in rows:
                app = {
                    'id': row['id'],
                    'name': row['name'],
                    'executable_path': row['executable_path'],
                    'icon_path': row['icon_path'],
                    'file_size': row['file_size'],
                    'file_type': row['file_type'],
                    'last_modified': row['last_modified'],
                    'launch_count': row['launch_count'],
                    'last_launched': row['last_launched'],
                    'is_favorite': bool(row['is_favorite']),
                    'created_at': row['created_at'],
                    'path_id': row['path_id'],
                    'folder_path': row['folder_path'],
                    'folder_name': row['folder_name']
                }
                apps.append(app)
            
            return apps
            
        except Exception as error:
            print(f"[DatabaseService] Ошибка получения избранных приложений: {error}")
            return []
    
    def get_apps_stats(self) -> Dict[str, int]:
        try:
            if not self.connection:
                print("[DatabaseService] Нет соединения с базой данных")
                return {
                    'total_apps': 0,
                    'total_paths': 0,
                    'total_launches': 0
                }
                
            cursor = self.connection.cursor()
            
            cursor.execute("""
                SELECT COUNT(*) as total_apps
                FROM apps a
                JOIN app_paths p ON a.path_id = p.id
                WHERE p.is_active = 1
            """)
            total_apps = cursor.fetchone()['total_apps']
            
            cursor.execute("SELECT COUNT(*) as total_paths FROM app_paths WHERE is_active = 1")
            total_paths = cursor.fetchone()['total_paths']
            
            cursor.execute("""
                SELECT COALESCE(SUM(launch_count), 0) as total_launches
                FROM apps a
                JOIN app_paths p ON a.path_id = p.id
                WHERE p.is_active = 1
            """)
            total_launches = cursor.fetchone()['total_launches']
            
            return {
                'total_apps': total_apps,
                'total_paths': total_paths,
                'total_launches': total_launches
            }
            
        except Exception as error:
            print(f"[DatabaseService] Ошибка получения статистики: {error}")
            return {
                'total_apps': 0,
                'total_paths': 0,
                'total_launches': 0
            }
    
    def get_app_by_path(self, executable_path: str) -> Optional[Dict[str, Any]]:
        try:
            if not self.connection:
                print("[DatabaseService] Нет соединения с базой данных")
                return None
                
            cursor = self.connection.cursor()
            
            query = """
                SELECT 
                    a.id,
                    a.name,
                    a.executable_path,
                    a.icon_path,
                    a.file_size,
                    a.file_type,
                    a.last_modified,
                    a.launch_count,
                    a.last_launched,
                    a.is_favorite,
                    a.created_at,
                    p.id as path_id,
                    p.path as folder_path,
                    p.name as folder_name
                FROM apps a
                JOIN app_paths p ON a.path_id = p.id
                WHERE p.is_active = 1 AND a.executable_path = ?
            """
            
            cursor.execute(query, (executable_path,))
            row = cursor.fetchone()
            
            if row:
                return {
                    'id': row['id'],
                    'name': row['name'],
                    'executable_path': row['executable_path'],
                    'icon_path': row['icon_path'],
                    'file_size': row['file_size'],
                    'file_type': row['file_type'],
                    'last_modified': row['last_modified'],
                    'launch_count': row['launch_count'],
                    'last_launched': row['last_launched'],
                    'is_favorite': bool(row['is_favorite']),
                    'created_at': row['created_at'],
                    'path_id': row['path_id'],
                    'folder_path': row['folder_path'],
                    'folder_name': row['folder_name']
                }
            
            return None
            
        except Exception as error:
            print(f"[DatabaseService] Ошибка поиска приложения по пути: {error}")
            return None
    
    def close(self):
        if self.connection:
            self.connection.close()
            self.connection = None
            print("[DatabaseService] Соединение с базой данных закрыто")

File: src/app/js/services/DatabaseService.js
import Database from 'better-sqlite3';
import path from 'path';
import fs from 'fs';
import paths from '../paths.js';

export class DatabaseService {
    static instance;
    
    constructor() {
        this.db = null;
        this.statements = {};
        this.initDatabase();
    }

    static getInstance() {
        if (!DatabaseService.instance) {
            DatabaseService.instance = new DatabaseService();
        }
        return DatabaseService.instance;
    }

    initDatabase() {
        try {
            const dbDir = paths.db_path;
            if (!fs.existsSync(dbDir)) {
                fs.mkdirSync(dbDir, { recursive: true });
            }

            const dbPath = path.join(dbDir, 'apps.db');
            this.db = new Database(dbPath);
            
            this.db.pragma('foreign_keys = ON');
            
            this.createTables();
            this.prepareStatements();
            
        } catch (error) {
            console.error('[DatabaseService] Ошибка инициализации:', error);
            throw new Error(`Ошибка инициализации базы данных: ${error.message}`);
        }
    }

    createTables() {
        const createTablesSQL = `
            CREATE TABLE IF NOT EXISTS app_paths (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                path TEXT UNIQUE NOT NULL,
                name TEXT NOT NULL,
                created_at DATETIME DEFAULT CURRENT_TIMESTAMP,
                last_scan DATETIME,
                is_active BOOLEAN DEFAULT 1
            );

            CREATE TABLE IF NOT EXISTS apps (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                path_id INTEGER NOT NULL,
                name TEXT NOT NULL,
                executable_path TEXT UNIQUE NOT NULL,
                icon_path TEXT,
                file_size INTEGER DEFAULT 0,
                file_type TEXT DEFAULT '.exe',
                last_modified DATETIME,
                launch_count INTEGER DEFAULT 0,
                last_launched DATETIME,
                created_at DATETIME DEFAULT CURRENT_TIMESTAMP,
                FOREIGN KEY (path_id) REFERENCES app_paths(id) ON DELETE CASCADE
            );

            CREATE INDEX IF NOT EXISTS idx_apps_name ON apps(name);
            CREATE INDEX IF NOT EXISTS idx_apps_launch_count ON apps(launch_count DESC);
            CREATE INDEX IF NOT EXISTS idx_apps_path_id ON apps(path_id);
            CREATE INDEX IF NOT EXISTS idx_app_paths_active ON app_paths(is_active);
        `;

        this.db.exec(createTablesSQL);
    }

    prepareStatements() {
        this.statements = {
            insertPath: this.db.prepare(`
                INSERT OR REPLACE INTO app_paths (path, name, is_active) 
                VALUES (?, ?, 1)
            `),
            
            reactivatePath: this.db.prepare(`
                UPDATE app_paths SET is_active = 1, last_scan = CURRENT_TIMESTAMP 
                WHERE path = ?
            `),
            
            getPathId: this.db.prepare(`
                SELECT id FROM app_paths WHERE path = ?
            `),
            
            getActivePathId: this.db.prepare(`
                SELECT id FROM app_paths WHERE path = ? AND is_active = 1
            `),
            
            updatePathScanTime: this.db.prepare(`
                UPDATE app_paths 
                SET last_scan = CURRENT_TIMESTAMP 
                WHERE id = ?
            `),
            
            getActivePaths: this.db.prepare(`
                SELECT * FROM app_paths 
                WHERE is_active = 1 
                ORDER BY name
            `),
            
            insertApp: this.db.prepare(`
                INSERT OR REPLACE INTO apps 
                (path_id, name, executable_path, file_size, file_type, last_modified) 
                VALUES (?, ?, ?, ?, ?, ?)
            `),
            
            deleteAppsByPath: this.db.prepare(`
                DELETE FROM apps WHERE path_id = ?
            `),
            
            getAllApps: this.db.prepare(`
                SELECT 
                    a.*,
                    p.path as folder_path,
                    p.name as folder_name
                FROM apps a
                JOIN app_paths p ON a.path_id = p.id
                WHERE p.is_active = 1
                ORDER BY a.launch_count DESC, a.name ASC
            `),

            getAppsByPathId: this.db.prepare(`
                SELECT * FROM apps WHERE path_id = ?
            `),
            
            searchApps: this.db.prepare(`
                SELECT 
                    a.*,
                    p.path as folder_path,
                    p.name as folder_name
                FROM apps a
                JOIN app_paths p ON a.path_id = p.id
                WHERE a.name LIKE ? AND p.is_active = 1
                ORDER BY a.launch_count DESC, a.name ASC
                LIMIT ?
            `),
            
            incrementLaunchCount: this.db.prepare(`
                UPDATE apps 
                SET launch_count = launch_count + 1,
                    last_launched = CURRENT_TIMESTAMP
                WHERE id = ?
            `),
            
            getAppById: this.db.prepare(`
                SELECT * FROM apps WHERE id = ?
            `),
            
            getAppStats: this.db.prepare(`
                SELECT 
                    COUNT(*) as total_apps,
                    COUNT(DISTINCT path_id) as total_paths,
                    SUM(launch_count) as total_launches
                FROM apps a
                JOIN app_paths p ON a.path_id = p.id
                WHERE p.is_active = 1
            `),
            
            deleteApp: this.db.prepare(`
                DELETE FROM apps WHERE id = ?
            `),
            
            deleteFolder: this.db.prepare(`
                UPDATE app_paths SET is_active = 0 WHERE id = ?
            `),
            
            deleteFolderApps: this.db.prepare(`
                DELETE FROM apps WHERE path_id = ?
            `)
        };
    }

    addPath(folderPath, folderName) {
        try {
            const result = this.statements.insertPath.run(folderPath, folderName);
            return result.changes > 0;
        } catch (error) {
            console.error('[DatabaseService] Ошибка добавления пути:', error);
            throw new Error(`Ошибка добавления пути: ${error.message}`);
        }
    }

    getPathId(folderPath) {
        try {
            const result = this.statements.getPathId.get(folderPath);
            return result ? result.id : null;
        } catch (error) {
            console.error('[DatabaseService] Ошибка получения ID пути:', error);
            return null;
        }
    }

    getActivePathId(folderPath) {
        try {
            const result = this.statements.getActivePathId.get(folderPath);
            return result ? result.id : null;
        } catch (error) {
            console.error('[DatabaseService] Ошибка получения активного ID пути:', error);
            return null;
        }
    }

    saveApps(folderPath, folderName, apps) {
        try {
            const transaction = this.db.transaction(() => {
                const activePathId = this.getActivePathId(folderPath);
                const anyPathId = this.getPathId(folderPath);
                
                if (activePathId) {
                    // Путь существует и активен - обновляем приложения
                    return this.updateAppsForPath(activePathId, apps);
                } else if (anyPathId) {
                    // Путь существует но неактивен - реактивируем и обновляем
                    this.statements.reactivatePath.run(folderPath);
                    return this.updateAppsForPath(anyPathId, apps);
                } else {
                    // Новый путь - создаем и добавляем все приложения
                    this.statements.insertPath.run(folderPath, folderName);
                    
                    const pathId = this.getPathId(folderPath);
                    if (!pathId) {
                        throw new Error('Не удалось получить ID пути');
                    }
                    
                    for (const app of apps) {
                        this.statements.insertApp.run(
                            pathId,
                            app.name,
                            app.path,
                            app.size || 0,
                            app.type || '.exe',
                            app.modified ? new Date(app.modified).toISOString() : new Date().toISOString()
                        );
                    }
                    
                    this.statements.updatePathScanTime.run(pathId);
                }
            });
            
            transaction();
            return true;
        } catch (error) {
            console.error('[DatabaseService] Ошибка сохранения приложений:', error);
            throw new Error(`Ошибка сохранения приложений: ${error.message}`);
        }
    }

    updateAppsForPath(pathId, apps) {
        try {
            const existingApps = this.statements.getAppsByPathId.all(pathId);
            const existingPaths = new Set(existingApps.map(app => app.path));
            
            let newAppsCount = 0;
            for (const app of apps) {
                if (!existingPaths.has(app.path)) {
                    this.statements.insertApp.run(
                        pathId,
                        app.name,
                        app.path,
                        app.size || 0,
                        app.type || '.exe',
                        app.modified ? new Date(app.modified).toISOString() : new Date().toISOString()
                    );
                    newAppsCount++;
                }
            }
            
            this.statements.updatePathScanTime.run(pathId);
            
            console.log(`[DatabaseService] Добавлено ${newAppsCount} новых приложений для существующего пути`);
            return true;
        } catch (error) {
            console.error('[DatabaseService] Ошибка обновления приложений:', error);
            throw error;
        }
    }

    getAppsForUI() {
        try {
            const apps = this.statements.getAllApps.all();
            
            const groupedApps = {};
            const paths = {};

            apps.forEach(app => {
                if (!paths[app.path_id]) {
                    paths[app.path_id] = {
                        id: app.path_id.toString(),
                        path: app.folder_path,
                        name: app.folder_name
                    };
                }

                if (!groupedApps[app.path_id]) {
                    groupedApps[app.path_id] = [];
                }

                groupedApps[app.path_id].push({
                    id: app.id.toString(),
                    name: app.name,
                    path: app.executable_path,
                    pathId: app.path_id.toString(),
                    launchCount: app.launch_count,
                    isFavorite: Boolean(app.is_favorite),
                    lastLaunched: app.last_launched
                });
            });

            return {
                paths: Object.values(paths),
                apps: groupedApps
            };
        } catch (error) {
            console.error('[DatabaseService] Ошибка получения приложений:', error);
            return { paths: [], apps: {} };
        }
    }

    searchApps(searchTerm, limit = 50) {
        try {
            return this.statements.searchApps.all(`%${searchTerm}%`, limit);
        } catch (error) {
            console.error('[DatabaseService] Ошибка поиска приложений:', error);
            return [];
        }
    }

    incrementLaunchCount(appId) {
        try {
            const result = this.statements.incrementLaunchCount.run(appId);
            return result.changes > 0;
        } catch (error) {
            console.error('[DatabaseService] Ошибка обновления счетчика запусков:', error);
            return false;
        }
    }

    getStats() {
        try {
            return this.statements.getAppStats.get();
        } catch (error) {
            console.error('[DatabaseService] Ошибка получения статистики:', error);
            return { total_apps: 0, total_paths: 0, total_launches: 0 };
        }
    }

    getActivePaths() {
        try {
            return this.statements.getActivePaths.all();
        } catch (error) {
            console.error('[DatabaseService] Ошибка получения путей:', error);
            return [];
        }
    }

    deleteApp(appId) {
        try {
            const result = this.statements.deleteApp.run(appId);
            return result.changes > 0;
        } catch (error) {
            console.error('[DatabaseService] Ошибка удаления приложения:', error);
            return false;
        }
    }

    deleteFolder(folderId) {
        try {
            this.statements.deleteFolderApps.run(folderId);
            
            const result = this.statements.deleteFolder.run(folderId);
            
            return result.changes > 0;
        } catch (error) {
            console.error('[DatabaseService] Ошибка удаления папки:', error);
            return false;
        }
    }

    getAppByPath(executablePath) {
        try {
            const result = this.statements.getAllApps.all().find(app => app.executable_path === executablePath);
            return result || null;
        } catch (error) {
            console.error('[DatabaseService] Ошибка поиска приложения по пути:', error);
            return null;
        }
    }

    close() {
        if (this.db) {
            this.db.close();
            this.db = null;
            console.log('[DatabaseService] Соединение с базой данных закрыто');
        }
    }
}
